src/status_im2/contexts/chat/messages/content/audio/view.cljs

```diff
@@ -0,0 +1,204 @@
+(ns status-im2.contexts.chat.messages.content.audio.view
+  (:require ["react-native-blob-util" :default ReactNativeBlobUtil]
+            [goog.string :as gstring]
+            [reagent.core :as reagent]
+            [react-native.audio-toolkit :as audio]
+            [status-im2.contexts.chat.messages.content.audio.style :as style]
+            [react-native.platform :as platform]
+            [taoensso.timbre :as log]
+            [quo2.foundations.colors :as colors]
+            [quo2.core :as quo]
+            [react-native.core :as rn]
+            [utils.re-frame :as rf]
+            [utils.i18n :as i18n]))
+
+(def ^:const media-server-uri-prefix "https://localhost:")
+(def ^:const audio-path "/messages/audio")
+(def ^:const uri-param "?messageId=")
+
+(defonce active-players (atom {}))
+(defonce audio-uris (atom {}))
+(defonce progress-timer (atom nil))
+(defonce current-player-key (reagent/atom nil))
+
+(defn get-player-key
+  [message-id in-pinned-view?]
+  (str in-pinned-view? message-id))
+
+(defn destroy-player
+  [player-key]
+  (when-let [player (@active-players player-key)]
+    (audio/destroy-player player)
+    (swap! active-players dissoc player-key)))
+
+(defn update-state
+  [state new-state]
+  (when-not (= @state new-state)
+    (reset! state new-state)))
+
+(defn seek-player
+  [player-key player-state value on-success]
+  (when-let [player (@active-players player-key)]
+    (audio/seek-player
+     player
+     value
+     #(when on-success (on-success))
+     #(update-state player-state :error))
+    (update-state player-state :seeking)))
+
+(defn download-audio-http
+  [base64-uri on-success]
+  (-> (.config ReactNativeBlobUtil (clj->js {:trusty platform/ios?}))
+      (.fetch "GET" (str base64-uri))
+      (.then #(on-success (.base64 ^js %)))
+      (.catch #(log/error "could not fetch audio " base64-uri))))
+
+(defn create-player
+  [{:keys [progress-ref player-state player-key]} audio-url on-success]
+  (download-audio-http
+   audio-url
+   (fn [base64-data]
+     (let [player (audio/new-player
+                   (str "data:audio/acc;base64," base64-data)
+                   {:autoDestroy                 false
+                    :continuesToPlayInBackground false}
+                   (fn []
+                     (update-state player-state :ready-to-play)
+                     (reset! progress-ref 0)
+                     (when (and @progress-timer (= @current-player-key player-key))
+                       (js/clearInterval @progress-timer)
+                       (reset! progress-timer nil))))]
+       (swap! active-players assoc player-key player)
+       (audio/prepare-player
+        player
+        #(when on-success (on-success))
+        #(update-state player-state :error)))))
+  (update-state player-state :preparing))
+
+(defn play-pause-player
```

Comment:
I see you tried to separate some of the state management outside the `f-audio-message` component, but still @briansztamfater, I'd just like to share some thoughts or suggestions.

1. A very good chunk of this namespace is not concerned or doesn't need to know about hiccup, reagent, etc, it's just raw timers and manipulation of atoms. In my experience in other React projects, this is a perfect candidate for a separate hook with good unit tests.
2. HTTP effects should happen at re-frame's event layer, ideally.
3. If we so desire to reduce usages of `use-effect`, we can also extract the whole state management into re-frame. Of course this is not a trivial task since the code is already written in the style of hooks, but it could lead to the best results.

But to summarize, I think the effectful part should exist as a separate hook, even outside this file, for example in `content/audio/hooks.cljs` and unit tested.

Reply:
@ilmotta As always, thanks for your valuable feedback, and agree with them. As I have another ongoing time consuming task and given that manual testing went fine, I wanted to merge this PR, and created an issue to keep track of this suggestion 👍  https://github.com/status-im/status-mobile/issues/15816
Will make sure to address it as soon as possible!